pkg/agent/flowexporter/connections/conntrack_linux.go

```diff
@@ -52,26 +50,34 @@ func NewConnTrackSystem(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet) *
 }
 
 // DumpFlows opens netlink connection and dumps all the flows in Antrea ZoneID of conntrack table.
-func (ct *connTrackSystem) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connection, error) {
+func (ct *connTrackSystem) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connection, *flowexporter.ConntrackOccupancy, error) {
 	// Get connection to netlink socket
 	err := ct.connTrack.Dial()
 	if err != nil {
-		klog.Errorf("Error when getting netlink socket: %v", err)
-		return nil, err
+		return nil, nil, fmt.Errorf("error when getting netlink socket: %v", err)
 	}
 
 	// ZoneID filter is not supported currently in tl-mo/conntrack library.
 	// Link to issue: https://github.com/ti-mo/conntrack/issues/23
 	// Dump all flows in the conntrack table for now.
 	conns, err := ct.connTrack.DumpFilter(conntrack.Filter{})
 	if err != nil {
-		klog.Errorf("Error when dumping flows from conntrack: %v", err)
-		return nil, err
+		return nil, nil, fmt.Errorf("error when dumping flows from conntrack: %v", err)
+	}
+
+	getMaxConnectionsInConnTrack := func() (int, error) {
+		return sysctl.GetSysctlNet("nf_conntrack_max")
+	}
+
+	maxConns, err := getMaxConnectionsInConnTrack()
```

Comment:
why not calling `sysctl.GetSysctlNet("nf_conntrack_max")` directly but creating the one line function?
And getting nf_conntrack_max seems not really related to the function, perhaps the function could return the filtered connections and all connections instead and use another method like `GetMaxConnections` to get `nf_conntrack_max` so that `DumpFlows` can focus on what it desires and no need to create a new struct `ConntrackOccupancy` just to pass the return value.

Reply:
I agree for the one line function. I think I suggested the helper function back when the function had some more logic.

Regarding adding a new `GetMaxConnections` method to the `ConnTrackDumper` interface, this is also my preferred solution. I suggested returning the values in `DumpFlows` to avoid updating metrics in multiple-places / platform-dependent code (https://github.com/vmware-tanzu/antrea/pull/1232#discussion_r488945704), but Quan's proposal is a nice improvement.

Reply:
Hi Quan, as @antoninbas mentioned before, creating a function is for symmetry with the userspace case. For nf_conntrack_max, since this value can be modify by admin anytime, so we decide to read max conn metric when reading the connection as a simple solution right now.

Reply:
Thanks for your advices, I have added GetMaxConnections as a new method for ConnTrackDumper interface and added unit tests for GetMaxConnections in kernel and user datapath respectively.